File: tests/rosetta/transpiler/Java/bitmap-histogram.java
public class Main {

    static int[][] image() {
        return new int[][]{new int[]{0, 0, 10000}, new int[]{65535, 65535, 65535}, new int[]{65535, 65535, 65535}};
    }

    static int[] histogram(int[][] g, int bins) {
        if (bins <= 0) {
            bins = g[0].length;
        }
        int[] h = new int[]{};
        int i = 0;
        while (i < bins) {
            h = java.util.stream.IntStream.concat(java.util.Arrays.stream(h), java.util.stream.IntStream.of(0)).toArray();
            i = i + 1;
        }
        int y = 0;
        while (y < g.length) {
            int[] row = g[y];
            int x = 0;
            while (x < row.length) {
                int p = row[x];
                int idx = ((Number)(((p * (bins - 1)) / 65535))).intValue();
h[idx] = h[idx] + 1;
                x = x + 1;
            }
            y = y + 1;
        }
        return h;
    }

    static int medianThreshold(int[] h) {
        int lb = 0;
        int ub = h.length - 1;
        int lSum = 0;
        int uSum = 0;
        while (lb <= ub) {
            if (lSum + h[lb] < uSum + h[ub]) {
                lSum = lSum + h[lb];
                lb = lb + 1;
            } else {
                uSum = uSum + h[ub];
                ub = ub - 1;
            }
        }
        return ((Number)(((ub * 65535) / h.length))).intValue();
    }

    static int[][] threshold(int[][] g, int t) {
        int[][] out = new int[][]{};
        int y_1 = 0;
        while (y_1 < g.length) {
            int[] row_1 = g[y_1];
            int[] newRow = new int[]{};
            int x_1 = 0;
            while (x_1 < row_1.length) {
                if (row_1[x_1] < t) {
                    newRow = java.util.stream.IntStream.concat(java.util.Arrays.stream(newRow), java.util.stream.IntStream.of(0)).toArray();
                } else {
                    newRow = java.util.stream.IntStream.concat(java.util.Arrays.stream(newRow), java.util.stream.IntStream.of(65535)).toArray();
                }
                x_1 = x_1 + 1;
            }
            out = appendObj(out, newRow);
            y_1 = y_1 + 1;
        }
        return out;
    }

    static void printImage(int[][] g) {
        int y_2 = 0;
        while (y_2 < g.length) {
            int[] row_2 = g[y_2];
            String line = "";
            int x_2 = 0;
            while (x_2 < row_2.length) {
                if (row_2[x_2] == 0) {
                    line = line + "0";
                } else {
                    line = line + "1";
                }
                x_2 = x_2 + 1;
            }
            System.out.println(line);
            y_2 = y_2 + 1;
        }
    }

    static void main() {
        int[][] img = image();
        int[] h_1 = histogram(img, 0);
        System.out.println("Histogram: " + _p(h_1));
        int t = medianThreshold(h_1);
        System.out.println("Threshold: " + _p(t));
        int[][] bw = threshold(img, t);
        printImage(bw);
    }
    public static void main(String[] args) {
        {
            long _benchStart = _now();
            long _benchMem = _mem();
            main();
            long _benchDuration = _now() - _benchStart;
            long _benchMemory = _mem() - _benchMem;
            System.out.println("{");
            System.out.println("  \"duration_us\": " + _benchDuration + ",");
            System.out.println("  \"memory_bytes\": " + _benchMemory + ",");
            System.out.println("  \"name\": \"main\"");
            System.out.println("}");
            return;
        }
    }

    static boolean _nowSeeded = false;
    static int _nowSeed;
    static int _now() {
        if (!_nowSeeded) {
            String s = System.getenv("MOCHI_NOW_SEED");
            if (s != null && !s.isEmpty()) {
                try { _nowSeed = Integer.parseInt(s); _nowSeeded = true; } catch (Exception e) {}
            }
        }
        if (_nowSeeded) {
            _nowSeed = (int)((_nowSeed * 1664525L + 1013904223) % 2147483647);
            return _nowSeed;
        }
        return (int)(System.nanoTime() / 1000);
    }

    static long _mem() {
        Runtime rt = Runtime.getRuntime();
        rt.gc();
        return rt.totalMemory() - rt.freeMemory();
    }

    static <T> T[] appendObj(T[] arr, T v) {
        T[] out = java.util.Arrays.copyOf(arr, arr.length + 1);
        out[arr.length] = v;
        return out;
    }

    static String _p(Object v) {
        return v != null ? String.valueOf(v) : "<nil>";
    }
}

File: tests/rosetta/transpiler/Kotlin/bitmap-histogram.kt
import java.math.BigInteger

var _nowSeed = 0L
var _nowSeeded = false
fun _now(): Long {
    if (!_nowSeeded) {
        System.getenv("MOCHI_NOW_SEED")?.toLongOrNull()?.let {
            _nowSeed = it
            _nowSeeded = true
        }
    }
    return if (_nowSeeded) {
        _nowSeed = (_nowSeed * 1664525 + 1013904223) % 2147483647
        kotlin.math.abs(_nowSeed)
    } else {
        kotlin.math.abs(System.nanoTime())
    }
}

fun toJson(v: Any?): String = when (v) {
    null -> "null"
    is String -> "\"" + v.replace("\"", "\\\"") + "\""
    is Boolean, is Number -> v.toString()
    is Map<*, *> -> v.entries.joinToString(prefix = "{", postfix = "}") { toJson(it.key.toString()) + ":" + toJson(it.value) }
    is Iterable<*> -> v.joinToString(prefix = "[", postfix = "]") { toJson(it) }
    else -> toJson(v.toString())
}

fun image(): MutableList<MutableList<Int>> {
    return mutableListOf(mutableListOf(0, 0, 10000), mutableListOf(65535, 65535, 65535), mutableListOf(65535, 65535, 65535))
}

fun histogram(g: MutableList<MutableList<Int>>, bins: Int): MutableList<Int> {
    var bins: Int = bins
    if (bins <= 0) {
        bins = (g[0]!!).size
    }
    var h: MutableList<Int> = mutableListOf<Int>()
    var i: Int = 0
    while (i < bins) {
        h = run { val _tmp = h.toMutableList(); _tmp.add(0); _tmp }
        i = i + 1
    }
    var y: Int = 0
    while (y < g.size) {
        var row: MutableList<Int> = g[y]!!
        var x: Int = 0
        while (x < row.size) {
            var p: Int = row[x]!!
            var idx: Int = (((p * (bins - 1)) / 65535).toInt())
            h[idx] = h[idx]!! + 1
            x = x + 1
        }
        y = y + 1
    }
    return h
}

fun medianThreshold(h: MutableList<Int>): Int {
    var lb: Int = 0
    var ub: BigInteger = (h.size - 1).toBigInteger()
    var lSum: Int = 0
    var uSum: Int = 0
    while ((lb).toBigInteger().compareTo((ub)) <= 0) {
        if ((lSum + h[lb]!!) < (uSum + h[(ub).toInt()]!!)) {
            lSum = lSum + h[lb]!!
            lb = lb + 1
        } else {
            uSum = uSum + h[(ub).toInt()]!!
            ub = ub.subtract((1).toBigInteger())
        }
    }
    return (((ub.multiply((65535).toBigInteger())).divide((h.size).toBigInteger())).toInt())
}

fun threshold(g: MutableList<MutableList<Int>>, t: Int): MutableList<MutableList<Int>> {
    var out: MutableList<MutableList<Int>> = mutableListOf<MutableList<Int>>()
    var y: Int = 0
    while (y < g.size) {
        var row: MutableList<Int> = g[y]!!
        var newRow: MutableList<Int> = mutableListOf<Int>()
        var x: Int = 0
        while (x < row.size) {
            if (row[x]!! < t) {
                newRow = run { val _tmp = newRow.toMutableList(); _tmp.add(0); _tmp }
            } else {
                newRow = run { val _tmp = newRow.toMutableList(); _tmp.add(65535); _tmp }
            }
            x = x + 1
        }
        out = run { val _tmp = out.toMutableList(); _tmp.add(newRow); _tmp }
        y = y + 1
    }
    return out
}

fun printImage(g: MutableList<MutableList<Int>>): Unit {
    var y: Int = 0
    while (y < g.size) {
        var row: MutableList<Int> = g[y]!!
        var line: String = ""
        var x: Int = 0
        while (x < row.size) {
            if (row[x]!! == 0) {
                line = line + "0"
            } else {
                line = line + "1"
            }
            x = x + 1
        }
        println(line)
        y = y + 1
    }
}

fun user_main(): Unit {
    var img: MutableList<MutableList<Int>> = image()
    var h: MutableList<Int> = histogram(img, 0)
    println("Histogram: " + h.toString())
    var t: Int = medianThreshold(h)
    println("Threshold: " + t.toString())
    var bw: MutableList<MutableList<Int>> = threshold(img, t)
    printImage(bw)
}

fun main() {
    run {
        System.gc()
        val _startMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _start = _now()
        user_main()
        System.gc()
        val _end = _now()
        val _endMem = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()
        val _durationUs = (_end - _start) / 1000
        val _memDiff = kotlin.math.abs(_endMem - _startMem)
        val _res = mapOf("duration_us" to _durationUs, "memory_bytes" to _memDiff, "name" to "main")
        println(toJson(_res))
    }
}
